File: src/main/kotlin/bedrockDragon/network/RakNet/peer/RakNetServerPeer.java
/*
 *    __     ______     ______     __  __     __   __     ______     ______  
 *   /\ \   /\  == \   /\  __ \   /\ \/ /    /\ "-.\ \   /\  ___\   /\__  _\
 *  _\_\ \  \ \  __<   \ \  __ \  \ \  _"-.  \ \ \-.  \  \ \  __\   \/_/\ \/  
 * /\_____\  \ \_\ \_\  \ \_\ \_\  \ \_\ \_\  \ \_\\"\_\  \ \_____\    \ \_\ 
 * \/_____/   \/_/ /_/   \/_/\/_/   \/_/\/_/   \/_/ \/_/   \/_____/     \/_/                                                                          
 *
 * the MIT License (MIT)
 *
 * Copyright (c) 2016-2020 "Whirvis" Trent Summerlin
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * the above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package bedrockDragon.network.RakNet.peer;

import static bedrockDragon.network.RakNet.RakNetPacket.*;

import java.net.InetSocketAddress;

import bedrockDragon.network.RakNet.RakNetPacket;
import bedrockDragon.network.RakNet.client.RakNetClient;
import bedrockDragon.network.RakNet.protocol.ConnectionType;
import bedrockDragon.network.RakNet.protocol.Reliability;
import bedrockDragon.network.RakNet.protocol.login.ConnectionRequestAccepted;
import bedrockDragon.network.RakNet.protocol.login.NewIncomingConnection;
import bedrockDragon.network.RakNet.protocol.message.EncapsulatedPacket;
import bedrockDragon.network.RakNet.protocol.message.acknowledge.Record;

import io.netty.channel.Channel;

/**
 * A server connection that handles login and other server related protocols.
 *
 * @author "Whirvis" Trent Summerlin
 * @since JRakNet v1.0.0
 */
public final class RakNetServerPeer extends RakNetPeer implements RakNetPeerMessenger {

	private final RakNetClient client;
	private EncapsulatedPacket loginRecord;
	private long timestamp;

	/**
	 * Creates a RakNet server peer.
	 * 
	 * @param client
	 *            the client that is connected to the server.
	 * @param address
	 *            the address of the peer.
	 * @param guid
	 *            the globally unique ID of the peer.
	 * @param maximumTransferUnit
	 *            the maximum transfer unit of the peer.
	 * @param connectionType
	 *            the connection type of the peer.
	 * @param channel
	 *            the channel to communicate to the peer with.
	 */
	public RakNetServerPeer(RakNetClient client, InetSocketAddress address, long guid, int maximumTransferUnit,
			ConnectionType connectionType, Channel channel) {
		super(address, guid, maximumTransferUnit, connectionType, channel);
		this.client = client;

		/*
		 * By the time this object is created, handshaking has begun between the
		 * server and client to finish login after connection.
		 */
		this.setState(RakNetState.HANDSHAKING);
	}

	@Override
	public long getTimestamp() {
		if (this.isLoggedIn()) {
			return System.currentTimeMillis() - timestamp;
		}
		return -1L;
	}

	@Override
	public void handleMessage(RakNetPacket packet, int channel) {
		if (packet.getId() == ID_CONNECTION_REQUEST_ACCEPTED && this.isHandshaking()) {
			ConnectionRequestAccepted connectionRequestAccepted = new ConnectionRequestAccepted(packet);
			connectionRequestAccepted.decode();
			if (!connectionRequestAccepted.failed()) {
				NewIncomingConnection newIncomingConnection = new NewIncomingConnection();
				newIncomingConnection.serverAddress = this.getAddress();
				newIncomingConnection.clientTimestamp = connectionRequestAccepted.clientTimestamp;
				newIncomingConnection.serverTimestamp = connectionRequestAccepted.serverTimestamp;
				newIncomingConnection.encode();
				if (!newIncomingConnection.failed()) {
					this.loginRecord = this.sendMessage(Reliability.RELIABLE_ORDERED_WITH_ACK_RECEIPT,
							newIncomingConnection);
				//	this.getLogger()
				//			.debug("Sent new incoming connection, waiting for acknowledgement before confirming login");
				} else {
					client.disconnect("Failed to login (" + newIncomingConnection.getClass().getSimpleName()
							+ " failed to encode)");
				}
			} else {
				client.disconnect("Failed to login (" + connectionRequestAccepted.getClass().getSimpleName()
						+ " failed to decode)");
			}
		} else if (packet.getId() == ID_DISCONNECTION_NOTIFICATION) {
			client.disconnect("Server disconnected");
		} else if (packet.getId() >= ID_USER_PACKET_ENUM) {
			client.callEvent(listener -> listener.handleMessage(client, this, packet, channel));
		} else {
			client.callEvent(listener -> listener.handleUnknownMessage(client, this, packet, channel));
		}
	}

	@Override
	public void onAcknowledge(Record record, EncapsulatedPacket packet) {
		if (record.equals(loginRecord.ackRecord)) {
			this.timestamp = System.currentTimeMillis();
			this.setState(RakNetState.LOGGED_IN);
			//this.getLogger().info("Logged in to server with globally unique ID "
			//		+ Long.toHexString(this.getGloballyUniqueId()).toUpperCase());
			client.callEvent(listener -> listener.onLogin(client, this));
		}
		client.callEvent(listener -> listener.onAcknowledge(client, this, record, packet));
	}

	@Override
	public void onNotAcknowledge(Record record, EncapsulatedPacket packet) {
		client.callEvent(listener -> listener.onLoss(client, this, record, packet));
	}

}

File: src/main/kotlin/bedrockDragon/debug/clientSimulator/RaknetServerPeer.kt
/*
 *    __     ______     ______     __  __     __   __     ______     ______
 *   /\ \   /\  == \   /\  __ \   /\ \/ /    /\ "-.\ \   /\  ___\   /\__  _\
 *  _\_\ \  \ \  __<   \ \  __ \  \ \  _"-.  \ \ \-.  \  \ \  __\   \/_/\ \/
 * /\_____\  \ \_\ \_\  \ \_\ \_\  \ \_\ \_\  \ \_\\"\_\  \ \_____\    \ \_\
 * \/_____/   \/_/ /_/   \/_/\/_/   \/_/\/_/   \/_/ \/_/   \/_____/     \/_/
 *
 * the MIT License (MIT)
 *
 * Copyright (c) 2016-2020 "Whirvis" Trent Summerlin
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * the above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package bedrockDragon.debug.clientSimulator

import bedrockDragon.network.raknet.protocol.packet.packethandler.logger
import bedrockDragon.network.raknet.InvalidChannelException
import bedrockDragon.network.raknet.RakNet
import bedrockDragon.network.raknet.RakNetPacket
import bedrockDragon.network.raknet.RakNetPacket.Companion.ID_CONNECTION_REQUEST_ACCEPTED
import bedrockDragon.network.raknet.RakNetPacket.Companion.ID_DISCONNECTION_NOTIFICATION
import bedrockDragon.network.raknet.RakNetPacket.Companion.ID_USER_PACKET_ENUM
import bedrockDragon.network.raknet.peer.RakNetPeer
import bedrockDragon.network.raknet.peer.Status
import bedrockDragon.network.raknet.protocol.ConnectionType
import bedrockDragon.network.raknet.protocol.Reliability
import bedrockDragon.network.raknet.protocol.login.ConnectionRequestAccepted
import bedrockDragon.network.raknet.protocol.login.NewIncomingConnection
import bedrockDragon.network.raknet.protocol.message.acknowledge.Record
import bedrockDragon.network.raknet.protocol.message.EncapsulatedPacket
import io.netty.channel.Channel
import java.net.InetSocketAddress
import java.util.concurrent.ConcurrentLinkedQueue

/**
 * A server connection that handles login and other server related protocols.
 *
 * @author "Whirvis" Trent Summerlin
 * @since JRakNet v1.0.0
 */
class RakNetServerPeer(
    private val client: RakNetClient,address: InetSocketAddress, guid: Long,maximumTransferUnit: Int,
    connectionType: ConnectionType, channel: Channel
) : RakNetPeer(address, guid, maximumTransferUnit, connectionType, channel){
    private var loginRecord: EncapsulatedPacket? = null
    private var timestamp: Long = 0
    private var messageIndex = 0
    var status : Status = Status.DISCONNECTED

    fun getTimestamp(): Long {
        return System.currentTimeMillis() - timestamp
    }

    fun handleMessage(packet: RakNetPacket, channel: Int) {
        if (packet.id == ID_CONNECTION_REQUEST_ACCEPTED) {
            val connectionRequestAccepted = ConnectionRequestAccepted(packet)
            connectionRequestAccepted.decode()
            if (!connectionRequestAccepted.failed()) {
                val newIncomingConnection = NewIncomingConnection()
                newIncomingConnection.serverAddress = this.address
                newIncomingConnection.clientTimestamp = connectionRequestAccepted.clientTimestamp
                newIncomingConnection.serverTimestamp = connectionRequestAccepted.serverTimestamp
                newIncomingConnection.encode()
                if (!newIncomingConnection.failed()) {
                    loginRecord = sendMessage(
                        Reliability.RELIABLE_ORDERED_WITH_ACK_RECEIPT,
                        newIncomingConnection
                    )
                } else {
                    client.disconnect(
                        "Failed to login (" + newIncomingConnection.javaClass.simpleName
                                + " failed to encode)"
                    )
                }
            } else {
                client.disconnect(
                    ("Failed to login (" + connectionRequestAccepted.javaClass.simpleName
                            + " failed to decode)")
                )
            }
        } else if (packet.id == ID_DISCONNECTION_NOTIFICATION) {
            client.disconnect("Server disconnected")
        } else if (packet.id >= ID_USER_PACKET_ENUM) {
            client.callEvent { listener: RakNetClientListener ->
                listener.handleMessage(
                    client, this, packet, channel
                )
            }
        } else {
            client.callEvent { listener: RakNetClientListener ->
                listener.handleUnknownMessage(
                    client, this, packet, channel
                )
            }
        }
    }

    fun sendMessage(reliability: Reliability, packet: RakNetPacket): EncapsulatedPacket? {
        return sendMessage(reliability, 0, packet)
    }

    fun sendMessage(reliability: Reliability, channel: Int, packet: RakNetPacket): EncapsulatedPacket {

            if (channel >= RakNet.CHANNEL_COUNT) {
                throw InvalidChannelException(channel);
            }

            // Generate encapsulated packet
            var encapsulated = EncapsulatedPacket();
            encapsulated.reliability = reliability;
            encapsulated.orderChannel = channel.toByte();
            encapsulated.payload = packet;
            if (reliability.isReliable) {
                encapsulated.messageIndex = this.bumpMessageIndex()
            }

            if (reliability.isOrdered or reliability.isSequenced) {
                encapsulated.orderIndex = if (reliability.isOrdered) orderSendIndex[channel]++ else
                sequenceSendIndex[channel]++;
            }

            // Add to send queue
            if (EncapsulatedPacket.needsSplit(this, encapsulated)) {
                encapsulated.splitId = ++splitId % 65536;
                for (split in EncapsulatedPacket.split(this, encapsulated)) {
                    sendQueue.add(split);
                }
            } else {
                sendQueue.add(encapsulated);
            }


            /*
             * Return a copy of the encapsulated packet as if a single variable is
             * modified in the encapsulated packet before it is sent, the
             * communication with the peer could cease to function entirely.
             */

            return encapsulated
    }

    fun onAcknowledge(record: Record, packet: EncapsulatedPacket?) {
        if ((record == loginRecord!!.ackRecord)) {
            timestamp = System.currentTimeMillis()

            client.callEvent { listener: RakNetClientListener ->
                listener.onLogin(
                    client, this
                )
            }
        }
        client.callEvent { listener: RakNetClientListener ->
            listener.onAcknowledge(
                client, this, record, packet
            )
        }
    }

    fun onNotAcknowledge(record: Record?, packet: EncapsulatedPacket?) {
        client.callEvent { listener: RakNetClientListener ->
            listener.onLoss(
                client, this, record, packet
            )
        }
    }

    fun update() {
        logger.info { "Connected" }
    }

    fun disconnect() {

    }

    fun handleInternal(packet: RakNetPacket) {

    }


}